File: Python/0153-minimum-rotated-sorted-array.py
class Solution:
    def findMin(self, nums: List[int]) -> int:
        left, right = 0, len(nums)-1
        target = nums[0]

        if nums[right] > nums[left]:
            return nums[0]

        while (left<right):
            mid = int(left + (right-left)/2)

            if (nums[mid] >= target):
                left = mid+1
            else:
                right = mid

        return nums[right]



#better
class Solution2:
    def findMin(self, nums: List[int]) -> int:
        left, right = 0, len(nums)-1

        while (left<right):
            mid = left + (right-left)//2

            if (nums[mid] < nums[right]):
                right = mid
            else:
                left = mid + 1

        return nums[left]

File: C++/0153-minimum-rotated-sorted-array.cpp
// Time O(log n)
// Space O(1)
class Solution {
public:
    int findMin(vector<int>& nums) {
        if(nums.size()==0) return -1;
        if(nums.size()==1) return nums[0];
        int left = 0, right = nums.size();
        int ans = -1;
        // search looks like
        // [T T T F F] (rotated) or 
        // [F F F F F] (already sorted)
        // we are looking for first false
        while (left <= right) {
            int mid = left + (right - left)/2;
            if (nums[mid] > nums[nums.size() - 1]) {
                left = mid + 1;
            } else {
                ans = mid;
                right = mid - 1;
            }
        }
        return nums[ans];
    }
};
